nxp_s32k/libuavcan/driver/include/s32k_libuavcan.hpp

```diff
@@ -0,0 +1,237 @@
+#include "libuavcan/platform/memory.hpp"
+
+/** STL queue for the intermediate ISR buffer */
+#include <deque>
+
+/** CMSIS Core for __REV macro use */
+#include "s32_core_cm4.h"
+
+/**
+ * Preprocessor conditionals for deducing the number of CANFD FlexCAN instances in target MCU,
+ * this macro is defined inside the desired memory map "S32K14x.h" included header file
+ */
+#if defined(MCU_S32K142) || defined(MCU_S32K144)
+#    define TARGET_S32K_CANFD_COUNT (1u)
+#    define DISCARD_COUNT_ARRAY 0
+
+#elif defined(MCU_S32K146)
+#    define TARGET_S32K_CANFD_COUNT (2u)
+#    define DISCARD_COUNT_ARRAY 0, 0
+
+#elif defined(MCU_S32K148)
+#    define TARGET_S32K_CANFD_COUNT (3u)
+#    define DISCARD_COUNT_ARRAY 0, 0, 0
+
+#else
+#    error "No NXP S32K compatible MCU header file included"
+#endif
+
+namespace libuavcan
+{
+namespace media
+{
+/**
+ * @namespace S32K
+ * Microcontroller-specific constants, variables and non-mutating helper functions for the use of the FlexCAN peripheral
+ */
+namespace S32K
+{
+/**
+ * Helper function for block polling a bit flag until it is set with a timeout of 0.2 seconds using a LPIT timer,
+ * the argument list and usage reassembles the classic block polling while loop, and instead of using a third
+ * argument to decide if it'ss a timed block for a clear or set, the two flavors of the function are provided.
+ *
+ * @param  flagRegister Register where the flag is located.
+ * @param  flagMask     Mask to AND'nd with the register for isolating the flag.
+ * @return libuavcan::Result::Success If the flag set before the timeout expiration..
+ * @return libuavcan::Result::Failure If a timeout ocurred before the desired flag set.
+ */
+libuavcan::Result flagPollTimeout_Set(volatile std::uint32_t& flagRegister, std::uint32_t flag_Mask);
+
+/**
+ * Helper function for block polling a bit flag until it is cleared with a timeout of 0.2 seconds using a LPIT timer
+ *
+ * @param  flagRegister Register where the flag is located.
+ * @param  flagMask     Mask to AND'nd with the register for isolating the flag.
+ * @return libuavcan::Result::Success If the flag cleared before the timeout expiration..
+ * @return libuavcan::Result::Failure If a timeout ocurred before the desired flag cleared.
+ */
+libuavcan::Result flagPollTimeout_Clear(volatile std::uint32_t& flagRegister, std::uint32_t flag_Mask);
+
+}  // END namespace S32K
+
+/**
+ * Implementation of the methods from libuavcan's media layer abstracct class InterfaceGroup,
+ * with the template arguments listed below; for further details of this interface class,
+ * refer to the template declaration in libuavcan/media/interface.hpp
+ *
+ * FrameT = Frame: MTUBytesParam = MaxFrameSizeBytes (64 bytes for CAN-FD)
+ *                 FlagBitsCompareMask = 0x00 (default)
+ * MaxTxFrames = 1 (default)
+ * MaxRxFrames = 1 (default)
+ */
+class S32K_InterfaceGroup : public InterfaceGroup<CAN::Frame<CAN::TypeFD::MaxFrameSizeBytes>>
+{
+private:
```

Comment:
The interface group should be abstract and have no private methods or data. Use aligned_storage and placement new to initialize concrete types from within the cpp file and return `S32K_InterfaceGroup` pointers from the interface manager. This will allow you to have a very simple header with no internals exposed to the consumer.

## Example of using placement new

```cpp
namespace
{
std::aligned_storage(sizeof(MyConcreteType), alignof(MyConcreteType)>::type s_my_type_storage;

} // end anonymous namespace
...

MyConcreteType* p = new(&s_my_type_storage) MyConcreteType(arguments);
```